crates/apub/src/objects/community.rs

```diff
@@ -1,101 +1,157 @@
 };
-use activitystreams_ext::Ext2;
-use anyhow::Context;
+use chrono::{DateTime, FixedOffset};
 use lemmy_api_common::blocking;
-use lemmy_db_queries::DbPool;
+use lemmy_apub_lib::{
+  values::{MediaTypeHtml, MediaTypeMarkdown},
+  verify_domains_match,
+};
+use lemmy_db_queries::{ApubObject, DbPool};
 use lemmy_db_schema::{
   naive_now,
   source::community::{Community, CommunityForm},
 };
-use lemmy_db_views_actor::community_moderator_view::CommunityModeratorView;
 use lemmy_utils::{
-  location_info,
-  utils::{check_slurs, check_slurs_opt, convert_datetime},
+  utils::{check_slurs, check_slurs_opt, convert_datetime, markdown_to_html},
   LemmyError,
 };
 use lemmy_websocket::LemmyContext;
+use serde::{Deserialize, Serialize};
+use serde_with::skip_serializing_none;
 use url::Url;
 
-#[async_trait::async_trait(?Send)]
-impl ToApub for Community {
-  type ApubType = GroupExt;
-
-  async fn to_apub(&self, pool: &DbPool) -> Result<GroupExt, LemmyError> {
-    let id = self.id;
-    let moderators = blocking(pool, move |conn| {
-      CommunityModeratorView::for_community(conn, id)
-    })
-    .await??;
-    let moderators: Vec<Url> = moderators
-      .into_iter()
-      .map(|m| m.moderator.actor_id.into_inner())
-      .collect();
-
-    let mut group = ApObject::new(Group::new());
-    group
-      .set_many_contexts(lemmy_context())
-      .set_id(self.actor_id.to_owned().into())
-      .set_name(self.title.to_owned())
-      .set_published(convert_datetime(self.published))
-      // NOTE: included attritubed_to field for compatibility with lemmy v0.9.9
-      .set_many_attributed_tos(moderators);
+#[skip_serializing_none]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Group {
+  #[serde(rename = "@context")]
+  context: OneOrMany<AnyBase>,
+  #[serde(rename = "type")]
+  kind: GroupType,
+  id: Url,
+  /// username, set at account creation and can never be changed
+  preferred_username: String,
+  /// title (can be changed at any time)
+  name: String,
+  content: Option<String>,
+  media_type: Option<MediaTypeHtml>,
+  source: Option<Source>,
+  icon: Option<ImageObject>,
+  /// banner
+  image: Option<ImageObject>,
+  // lemmy extension
+  sensitive: Option<bool>,
+  // lemmy extension
+  pub(crate) moderators: Option<Url>,
+  inbox: Url,
+  pub(crate) outbox: Url,
+  followers: Url,
+  endpoints: Endpoints<Url>,
+  public_key: PublicKey,
+  published: DateTime<FixedOffset>,
+  updated: Option<DateTime<FixedOffset>>,
+  #[serde(flatten)]
+  unparsed: Unparsed,
+}
 
-    if let Some(u) = self.updated.to_owned() {
-      group.set_updated(convert_datetime(u));
-    }
-    if let Some(d) = self.description.to_owned() {
-      set_content_and_source(&mut group, &d)?;
-    }
+impl Group {
+  pub(crate) fn id(&self, expected_domain: &Url) -> Result<&Url, LemmyError> {
+    verify_domains_match(&self.id, expected_domain)?;
+    Ok(&self.id)
+  }
+  pub(crate) async fn from_apub_to_form(
```

Comment:
Seems like this should be a trait function with associated types, if its done for the different things ( Group, Person, Note, etc ). Just like the `ToApub` / `FromApub`

Reply:
This is only implemented for Group now, and not for the other types. Its needed to handle the Update/Community activity.